### PR TITLE
A generic mechnism for supporting new type

### PR DESCRIPTION
**Pull request by hujun-open**
after last PR, I unfortunately realized that there are a few more custom struct to support in my project, so I think there should be a mechanism for user to add support themself, and also the way last PR work could be better, so this is what this PR does, it makes supporting new type much easier 

I rewrite implementation for time.Time, net.IP, net.IPNet, and net.HardwareAddr with the new way. 

however this PR uses golang generic mechanism, which requires go ver1.18+, I uses 1.19 in go.mod and CI, it seems you want to use 1.16, so if you don't want take this PR, it is fine, I will just use my fork

### CODE REVIEW

**Comment by itzg** [comment]
Yeah, after realizing the version change "broke" 3 or 4 of my applications I downgraded since it wasn't super important to use 1.20. 

I really like the concept of your latest change, so let me give some  thought about going to a v2 branch or just going with 1.10+ requiring  1.20. Any suggestions on either of those? (I can "fix" my applications by telling dependabot to ignore the version change)

**Comment by hujun-open** [comment]
if you plan to maintain both v1 and v2 branch, then create a v2 branch; otherwise just bump the minor version

**Comment by itzg** [comment]
Yeah, bumping the minor version does seem best to me. 1.10.x and 1.11.0 can pick up these newest changes.

File: simple.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func (v *simpleType[T]) SetRef(t *T) {
+	v.val = t
+}
+
+type handlerFunc func(tag reflect.StructTag, fieldRef interface{},
+	hasDefaultTag bool, tagDefault string,
+	flagSet *flag.FlagSet, renamed string,
+	usage string, aliases string) error
```
Comment (itzg) [inline simple.go:58]: I think it would be better to declare this in `general.go` alongside `supportedStructList`

File: general.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"flag"
+	"fmt"
+	"strings"
+)
+
+// this is a list of supported struct, like time.Time, that walkFields() won't walk into,
+// the key is the is string returned by the getTypeName(<struct_type>),
+// each supported struct need to be added in this map in init()
+var supportedStructList = make(map[string]handlerFunc)
```
Comment (itzg) [inline general.go:16]: This needs a new name. Something like `extendedTypes`?

**Review by itzg — CHANGES_REQUESTED** [review]
Just the one little change and it's good to go. Thanks!

File: general.go (hunk 2 above)
Comment (hujun-open) [reply general.go]: > Just the one little change and it's good to go. Thanks!

done, also updated readme

**Review by itzg — APPROVED** [review]

**Comment by itzg** [comment]
Released https://github.com/itzg/go-flagsfiller/releases/tag/v1.11.0

**Pull request merged**